Allow creating IPC::Connection with a SerialFunctionDispatcher other than the main RunLoop.

mattwoodrow opened the pull request:
#### 4ba481e2960272016ebb1bd6f6132a08a6f8eb67
<pre>
Allow creating IPC::Connection with a SerialFunctionDispatcher other than the main RunLoop.
<a href="https://bugs.webkit.org/show_bug.cgi?id=244825">https://bugs.webkit.org/show_bug.cgi?id=244825</a>

Reviewed by Dean Jackson.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::SyncMessageStateImpl::enqueueMatchingMessages):
(IPC::Connection::SyncMessageStateImpl::processIncomingMessage):
(IPC::Connection::SyncMessageStateImpl::dispatchMessages):
(IPC::Connection::SyncMessageStateImpl::dispatchMessagesAndResetDidScheduleDispatchMessagesForConnection):
(IPC::Connection::createServerConnection):
(IPC::Connection::Connection):
(IPC::Connection::~Connection):
(IPC::Connection::connection):
(IPC::Connection::enqueueMatchingMessagesToMessageReceiveQueue):
(IPC::Connection::invalidate):
(IPC::Connection::waitForMessage):
(IPC::Connection::sendSyncMessage):
(IPC::Connection::processIncomingMessage):
(IPC::Connection::enableIncomingMessagesThrottling):
(IPC::Connection::connectionDidClose):
(IPC::Connection::dispatchSyncMessage):
(IPC::Connection::dispatchDidReceiveInvalidMessage):
(IPC::Connection::enqueueIncomingMessage):
(IPC::Connection::dispatchMessage):
(IPC::Connection::dispatchIncomingMessages):
(IPC::Connection::wakeUpRunLoop):
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::sendSync):
(IPC::Connection::waitForAndDispatchImmediately):
(IPC::Connection::waitForAsyncCallbackAndDispatchImmediately):

Canonical link: <a href="https://commits.webkit.org/255868@main">https://commits.webkit.org/255868@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a5404f0349ca4d3068e83f1ad34c05abeccd1a19

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93857 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3049 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24428 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103489 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3065 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31284 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86177 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99494 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99524 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2186 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80280 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29207 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/84112 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83814 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72159 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37680 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/17641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35544 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18904 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4051 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39423 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41358 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38134 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WTF/wtf/FunctionDispatcher.h:
```diff
@@ -42,6 +44,23 @@ class FunctionDispatcher {
     WTF_EXPORT_PRIVATE FunctionDispatcher();
 };
 
+class WTF_CAPABILITY("is current") SerialDispatchQueue : public FunctionDispatcher, public ThreadSafeRefCounted<SerialDispatchQueue> {
```
SerialDispatchQueue is not consistent naming, if we have FunctionDispatchers and WorkQueues. So it would be SerialFunctionDispatcher.

If you want to go with this naming, you would probably move the FunctionDispatcher from the WorkQueueBase to ConcurrentWorkQueue and then  add SerialFunctionDispatcher to WorkQueue.

Also SerialFunctionDispatcher cannot be a refcounted, as there might be useful serial function dispatchers that are not refcounted.

I don't think SerialFunctionDispatcher is needed at the moment.
I would rather see just Connection use FunctionDispatcher and document that it cannot, for the time being, be a concurrent one.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -286,13 +316,14 @@ static HashMap<uintptr_t, HashMap<uint64_t, CompletionHandler<void(Decoder*)>>>&
+    , m_dispatcher(dispatcher ? *dispatcher : RunLoop::current())
 {
-    ASSERT(RunLoop::isMain());
+    assertIsCurrent(m_dispatcher);
```
This assert is not needed, e.g. it's not needed that we are currently in the dispatcher sequence when constructing the connection.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -342,9 +369,10 @@ void Connection::setShouldExitOnSyncMessageSendFailure(bool shouldExitOnSyncMess
 void Connection::enqueueMatchingMessagesToMessageReceiveQueue(MessageReceiveQueue& receiveQueue, const ReceiverMatcher& receiverMatcher)
 {
-    ASSERT(isMainRunLoop());
+    assertIsCurrent(m_dispatcher);
```
For these sequential access assertions, you could do:

```
Locker lock { m_sequentialAccess };

Connection {
  NO_UNIQUE_ADDRESS SequetialAccessLock;
}
```

This could be done in a separate patch that also introduces WTF::SequentialAccessLock. (The lock is just a dummy that asserts that threadid == locked thread id.)

mattwoodrow replied:
Isn't that a considerably weaker guarantee than the current assertions?

A consumer could use a single connection from multiple threads (without external synchronisation), and depending on timing might not hit the assertions during testing. The code would still be broken though, and could cause issues with different workloads/timings.

What's the benefit of allowing this 'non-sequential' pattern, rather than being permanently associated with a given thread (or SerialFunctionDispatcher thing)?

kkinnunen-apple replied:
> Isn't that a considerably weaker guarantee than the current assertions?

Yes. In general we want the weakest possible guarantees that satisfies the requirements of the system. This gives the widest possible uses and widest possible future modification directions.

So we need to meticulously understand what the system requires.

> A consumer could use a single connection from multiple threads (without external synchronisation), and depending on timing might not hit the assertions during testing. The code would still be broken though, and could cause issues with different workloads/timings.

This is a valid argument.
 
> What's the benefit of allowing this 'non-sequential' pattern, rather than being permanently associated with a given thread (or SerialFunctionDispatcher thing)?

I explained this in the other comment.
Unless needed, I'd be hesitant to add the semantics of "SerialFunctionDispatcher either is current or is not current". This is because then all SerialFunctionDispatchers would need that property to apply to the functions dispatched through them. This would (possibly?) mean that the functions dispatched through them would need to carry this additional state even in cases it would not be carried.

If you think it'd be easier, maybe it's beneficial to add this to FunctionDispatcher / SerialFunctionDispatcher and be done with it, and then I can think if it's really a problem later.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -1023,7 +1071,14 @@ void Connection::dispatchSyncMessage(Decoder& decoder)
 void Connection::dispatchDidReceiveInvalidMessage(MessageName messageName)
 {
-    ensureOnMainRunLoop([this, protectedThis = Ref { *this }, messageName]() mutable {
+    if (isMainRunLoopConnection() && isMainRunLoop()) {
```
This is a tricky sequence.
My initial thought is that the caller should ensure that the dispatchDidReceiveInvalidMessage is called in the IPC receive work queue. Currently the only caller dispatches, which is silly for Cocoa. I think also Unix runs in the receive work queue, not sure about windows. It *should*.

This could be done in a separate patch and tested in itself.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -129,6 +129,34 @@ class Connection::SyncMessageState {
     Deque<ConnectionAndIncomingMessage> m_messagesToDispatchWhileWaitingForSyncReply WTF_GUARDED_BY_LOCK(m_lock);
 };
 
+// Connections created for background threads don't (yet) support receiving messages while waiting for a sync reply.
```
I don't think this is a the best approach, it makes it overly complicated to understand if it's correct or not and overly complicated to understand how the sync messages and waited on messages are dispatched.

The wait here, I think, completes before the message has been dispatched. This is not what the system is trying to achieve.

I think better is to:
1. Make a ReentrantFunctionDispatcher that supports dispatchSync()
2. Make a mapping of ReentrantFunctionDispatcher -> SyncMessageState

Other alternative is to ensure Non-mainthread connections do not have non-complex received messages. E.g. RRBProxy could wait for a sync point for flush to complete, but then the "did create backend" message would need another solution.  This is not very general solution.

This, too, should be worked on separately and tests should be added.

kkinnunen-apple reviewed (commented):
In general:
There is now TestIPC test runner, so you can add a test testing each change that is not related to assertions.

The changes here are complex, and they could be done more gradually. It is easy to make a behaviour difference here.

I would suggest that this means isolating following aspects:
1. asserting that the client calls the IPC::Connection correctly
2. running message handlers in arbitrary, user selected function dispatcher
3. sending async messages from a thread
4. sending sync messages from a thread
5. waiting for sync messages

In general I do not think it's a good idea to incorporate "is current" concept to the function dispatcher. Not all useful function dispatchers maintain such a state, and it would be shame to require them to do so. You can imagine a function dispatcher that wraps the closure in its own closure and sends it to another function dispatcher, say main run loop. In this case, it would be shame to require yet-another wrapping that would then set the is current state. It's also not required for this change.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -196,6 +196,8 @@ class Connection : public ThreadSafeRefCounted<Connection, WTF::DestructionThrea
     static Connection* connection(UniqueID);
     UniqueID uniqueID() const { return m_uniqueID; }
 
+    SerialDispatchQueue* dispatcher() { return m_dispatcher.ptr(); }
```
This is unused function

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -359,6 +362,10 @@ class Connection : public ThreadSafeRefCounted<Connection, WTF::DestructionThrea
 
     Timeout timeoutRespectingIgnoreTimeoutsForTesting(Timeout) const;
 
+    virtual bool isMainRunLoopConnection() { return true; }
```
I don't think these need virtual functions
(and the solution probably should be something different than BackgroundConnection anyway)

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -547,7 +556,7 @@ uint64_t Connection::sendWithAsyncReply(T&& message, C&& completionHandler, uint
         else
             T::cancelReply(WTFMove(completionHandler));
-    }, CompletionHandlerCallThread::MainThread));
+    }));
```
This might be a behavioural difference in some esoteric current call sites, so care should be taken to change this. E.g add an local assert that all callers are in main thread, run layout tests, then make a self-contained change removing this parameter explaining that this was indeed a redundant parameter here.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -75,11 +75,22 @@ struct Connection::WaitForMessageState {
-    ~SyncMessageState() = delete;
+    }
+    ~SyncMessageStateImpl() {
+        Locker locker { s_syncStateLock };
```
Locking inside a destructor almost never works. You lock because many threads might reference the same object. But you cannot have  multiple references to an object that is being destroyed, that's the antithesis of destruction!

In this case it is incorrect.
So here you are already destroying the object, but pointer to it is in the map. So other thread might come and grab that pointer from the map, take a ref (from zero to one!!) and continue using it. Then this thread would continue the destructor, carrying out the destruction and removal from the map.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -101,13 +112,22 @@ class Connection::SyncMessageState {
 private:
+    static HashMap<SerialFunctionDispatcher*, SyncMessageStateImpl*>& syncStateMap() WTF_REQUIRES_LOCK(s_syncStateLock)
+    {
+        static NeverDestroyed<HashMap<SerialFunctionDispatcher*, SyncMessageStateImpl*>> map;
```
neverdestroyed is not thread safe because we don't have thread-safe statics

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -291,6 +319,7 @@ Connection::Connection(Identifier identifier, bool isServer, Client& client)
     , m_uniqueID(UniqueID::generate())
     , m_isServer(isServer)
     , m_connectionQueue(WorkQueue::create("com.apple.IPC.ReceiveQueue"))
+    , m_syncState(SyncMessageStateImpl::forCurrentThread(&RunLoop::main()))
```
the sync state association must happen on open().

E.g. the connection might be created on main run loop, but the open says "deliver to this message receiver." So here you then want to change the api contract to "deliver to this message receiver in this runloop".

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -1005,7 +1034,7 @@ void Connection::dispatchSyncMessage(Decoder& decoder)
-        SyncMessageState::singleton().dispatchMessages();
+        syncState().dispatchMessages();
     } else {
         // Hand off both the decoder and encoder to the client.
```
in Connection::invalidate() you need to add a note 
// Note: currently we bind the sync state to main runloop, which will always exist. As such, we do not yet disconnect the connection from it, even though in reality we should not dispatch through the runloop after invalidate has happened.

kkinnunen-apple replied:
But then you might note that you need the explicit release operation.
Then you have the option where should that happen: in the wrong place (~Connection) or in the right place (Connection::invalidate). If you then choose the right place, then you need to prove that the ptr is not accessed in  a wrong way from receive thread nor from the submitted closures. Which then leads to some other changes..

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -418,14 +418,21 @@ class Connection : public ThreadSafeRefCounted<Connection, WTF::DestructionThrea
     struct PendingSyncReply;
     Vector<PendingSyncReply> m_pendingSyncReplies WTF_GUARDED_BY_LOCK(m_syncReplyStateLock);
 
+    Ref<SyncMessageState> m_syncState;
```
Most likely this cannot be a refcounted, so you end up with SyncMessageState* and then you don't need the Impl rename in the .cpp.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -75,11 +75,22 @@ struct Connection::WaitForMessageState {
+class Connection::SyncMessageStateImpl : public Connection::SyncMessageState {
 public:
-    static SyncMessageState& singleton();
+    static Ref<Connection::SyncMessageState> forCurrentThread(SerialFunctionDispatcher*);
```
Your lookup function cannot be
"SyncMessageState for current thread serial function dispatcher".
Name doesn't make sense, it does not look anything up for current thread.

So the lookup function is 
"SyncMessageState for this serial function dispatcher."
So I'd name it "SyncMessageState::get(SerialFunctionDispatcher&)".

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -171,7 +171,7 @@ Ref<Connection::SyncMessageState> Connection::SyncMessageStateImpl::forCurrentTh
 
 void Connection::SyncMessageStateImpl::enqueueMatchingMessages(Connection& connection, MessageReceiveQueue& receiveQueue, const ReceiverMatcher& receiverMatcher)
```
So these changes are testable in TestIPC, so you can create bunch of message sending tests that start new runloops and send messages back and forth.
The tests could test new runloops in different threads and in same threads.

kkinnunen-apple replied:
Another test set needs to be such that you write a mock serialfunctiondispatcher that detects whether or not it is accessed incorrectly.
E.g. one thread sends incorrect messages
the receiver marks the connection invalid at time point a.
the receiver uses the mock function dispatcher to detect that the function dispatcher was not accessed after invalidate().

E.g. before this patch it didn't make a difference, because the thing only used main runloop which is always around. Now the function dispatcher might go away, and as such the access semantics must be precise (which they are not until the bugs are fixed)..

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -314,6 +367,14 @@ Connection::~Connection()
     clearAsyncReplyHandlers(*this);
 }
 
+void Connection::openBase(Client& client)
```
Having pingpong of 

- invalidate() in base implementation, platformInvalidate() in platform implementation
- open() in platform implementation, openBase() in base implementation 

is confusing.

It's better to have just one pattern, e.g. 
```
void Thing::somethingSomething() {
        platformSomethingSomething()
        .....
}
```

So this could be:
```
bool Connection::open(Client& client)
{
     if (!platformPrepareForOpen())
           return false;
     m_client = &client;
     m_syncState  = SyncMessageState::get(RunLoop::main());
     platformOpen();
     return true;
}
```

mattwoodrow replied:
Sure, I can swap these around.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -341,9 +402,9 @@ void Connection::setShouldExitOnSyncMessageSendFailure(bool shouldExitOnSyncMess
 void Connection::enqueueMatchingMessagesToMessageReceiveQueue(MessageReceiveQueue& receiveQueue, const ReceiverMatcher& receiverMatcher)
 {
-    ASSERT(isMainRunLoop());
+    ASSERT_MAIN_RUN_LOOP_AND_SYNC_STATE;
```
Let's not add convoluted macro-based solutions if possible.
This should be
`assertIsCurrent(dispatcher());`

mattwoodrow replied:
I got rid of the `dispatcher()` function, since it'd have to be `WTF_REQUIRES_LOCK` (since the dispatcher pointer can go away at any time, if you're not on the dispatcher thread).

That means this would have to be:

`{
    Locker lock { m_syncStateLock; }
    assertIsCurrent(dispatcher());
}`

For each of the places we check it. I realise that this is just an assert, and the dispatcher can't racily go away (assuming that the code is correct, and this assert always succeeds), but I don't think TSA will (or should) let us ignore that.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -127,18 +152,45 @@ class Connection::SyncMessageState {
-Connection::SyncMessageState& Connection::SyncMessageState::singleton()
+Lock Connection::SyncMessageState::s_syncStateLock;
+
+DataMutexLocker<RefPtr<Connection::SyncMessageState>> Connection::syncState()
```
So this DataMutex is just a convolution.
So you need to decide:
- if the reference to SyncMessageState is released in ~Connection(), then there's nothing to be locked.
- if the reference to SyncMessageState is released in Connection::invalidate(), then the incomingMessagesLock should protect the m_syncState.

mattwoodrow replied:
I think we agreed that it has to be released in `invalidate()`, since we need the `SerialFunctionDispatcher` pointer to go away entirely during `invalidate()` since it's not refcounted and the consumer of Connection needs to be able to immediately remove all references to it.

I'm not super convinced that sharing mutexes for more data members is simpler (especially since DataMutex mostly hides the mutex and just enforces compile-time locked access), but happy to make that change if you prefer.

kkinnunen-apple reviewed (commented):
I think it's a bit too complicated:
- Adding redundant locking is more confusing than not.
- The thread-safe map lookup and reference counting is a bit convoluted (for my taste?).

There's a lot of (already) overlapping locks in Connection to be fixed, so we should not just add redundant ones (the data locker thing).

I added some suggestions but there's some work to be done realising these.

In my opinion this patch could be worked on as a separate issue for simplicity of review and getting bots to compile it. Either by working on the code structure with RunLoop or then first submitting the previous patches.

grorg reviewed (approved):
I'll r+ this patch, even though it is now significantly different from what is discussed in all the comments above. It seems they will be addressed later and this is just the smallest possible change required to unblock the offscreen canvas work.

I hope this is ok.

mattwoodrow committed:
Allow creating IPC::Connection with a SerialFunctionDispatcher other than the main RunLoop.
https://bugs.webkit.org/show_bug.cgi?id=244825

Reviewed by Dean Jackson.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::SyncMessageStateImpl::enqueueMatchingMessages):
(IPC::Connection::SyncMessageStateImpl::processIncomingMessage):
(IPC::Connection::SyncMessageStateImpl::dispatchMessages):
(IPC::Connection::SyncMessageStateImpl::dispatchMessagesAndResetDidScheduleDispatchMessagesForConnection):
(IPC::Connection::createServerConnection):
(IPC::Connection::Connection):
(IPC::Connection::~Connection):
(IPC::Connection::connection):
(IPC::Connection::enqueueMatchingMessagesToMessageReceiveQueue):
(IPC::Connection::invalidate):
(IPC::Connection::waitForMessage):
(IPC::Connection::sendSyncMessage):
(IPC::Connection::processIncomingMessage):
(IPC::Connection::enableIncomingMessagesThrottling):
(IPC::Connection::connectionDidClose):
(IPC::Connection::dispatchSyncMessage):
(IPC::Connection::dispatchDidReceiveInvalidMessage):
(IPC::Connection::enqueueIncomingMessage):
(IPC::Connection::dispatchMessage):
(IPC::Connection::dispatchIncomingMessages):
(IPC::Connection::wakeUpRunLoop):
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::sendSync):
(IPC::Connection::waitForAndDispatchImmediately):
(IPC::Connection::waitForAsyncCallbackAndDispatchImmediately):

Canonical link: https://commits.webkit.org/255868@main
